pkg/app/piped/planner/kubernetes/kubernetes.go

```diff
@@ -480,3 +492,43 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
 
 	return b.String(), nil
 }
+
+// determineVersions decides artifact versions of an application.
+// It finds all container images that are being specified in the workload manifests then returns their names, version numbers, and urls.
+func determineVersions(manifests []provider.Manifest) ([]*model.ArtifactVersion, error) {
+	versions := []*model.ArtifactVersion{}
+	imageMap := map[string]struct{}{}
+
+	for _, m := range manifests {
+		// TODO: Determine container image version from other workload kinds such as StatefulSet, Pod, Daemon, CronJob...
+		if !m.Key.IsDeployment() {
+			continue
+		}
+		data, err := m.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		var d resource.Deployment
+		if err := json.Unmarshal(data, &d); err != nil {
+			return nil, err
+		}
+
+		containers := d.Spec.Template.Spec.Containers
+		// remove duplicate images on multiple manifests
+		for _, c := range containers {
+			imageMap[c.Image] = struct{}{}
+		}
+	}
+
```

Comment:
Move initialization versions slice to later to know its size.

```suggestion
	imageMap := map[string]struct{}{}

	for _, m := range manifests {
		// TODO: Determine container image version from other workload kinds such as StatefulSet, Pod, Daemon, CronJob...
		if !m.Key.IsDeployment() {
			continue
		}
		data, err := m.MarshalJSON()
		if err != nil {
			return nil, err
		}
		var d resource.Deployment
		if err := json.Unmarshal(data, &d); err != nil {
			return nil, err
		}

		containers := d.Spec.Template.Spec.Containers
		// remove duplicate images on multiple manifests
		for _, c := range containers {
			imageMap[c.Image] = struct{}{}
		}
	}
	
	versions := make([]*model.ArtifactVersion, 0, len(imageMap))
```

Reply:
thank you for suggestion!